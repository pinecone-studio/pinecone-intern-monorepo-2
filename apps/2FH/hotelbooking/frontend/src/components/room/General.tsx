'use client';

import { useState } from 'react';
import { Generalmode } from './';

interface FormData {
  name: string;
  type: string[];
  pricePerNight: string;
  roomInformation: string[];
  bedNumber: number;
}

interface GeneralProps {
  onSave: (_data: FormData) => void;
  loading: boolean;
  _onImageSave: (_images: string[]) => void;
  _data: FormData;
}

export const General = ({ onSave, loading, _onImageSave, _data }: GeneralProps) => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  const handleOpenModal = () => {
    setIsModalOpen(true);
  };

  const handleCloseModal = () => {
    setIsModalOpen(false);
  };

  const handleSave = (data: FormData) => {
    onSave(data);
    setIsModalOpen(false);
  };

  const formatValue = (value: string | string[]) => {
    if (Array.isArray(value)) {
      return value.length > 0 ? value.join(', ') : '-/-';
    }
    return value || '-/-';
  };

  return (
    <div data-cy="General" className="flex flex-col gap-y-4">
      <div className="flex justify-between items-center">
        <h2 data-cy="General-Info" className="text-lg font-semibold text-gray-900">
          General Info
        </h2>
        <button data-cy="Edit-General" className="px-4 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition-colors text-sm font-medium" onClick={handleOpenModal}>
          Edit
        </button>
      </div>

      <div className="space-y-3">
        <div className="flex justify-between items-center py-2">
          <span className="text-sm font-medium text-gray-700">Name:</span>
          <span data-cy="General-Name-Value" className="text-sm text-gray-900">
            {formatValue(_data.name)}
          </span>
        </div>
        <div className="flex justify-between items-center py-2">
          <span className="text-sm font-medium text-gray-700">Type:</span>
          <span className="text-sm text-gray-900">{formatValue(_data.type)}</span>
        </div>
        <div className="flex justify-between items-center py-2">
          <span className="text-sm font-medium text-gray-700">Price per night:</span>
          <span className="text-sm text-gray-900">{formatValue(_data.pricePerNight)}</span>
        </div>
        <div className="flex justify-between items-center py-2">
          <span className="text-sm font-medium text-gray-700">Room Information:</span>
          <span className="text-sm text-gray-900">{formatValue(_data.roomInformation)}</span>
        </div>
        <div className="flex justify-between items-center py-2">
          <span className="text-sm font-medium text-gray-700">Bed Number:</span>
          <span className="text-sm text-gray-900">{_data.bedNumber || '-/-'}</span>
        </div>
      </div>

      <Generalmode isOpen={isModalOpen} onClose={handleCloseModal} onSave={handleSave} loading={loading} initialData={_data} />
    </div>
  );
};
