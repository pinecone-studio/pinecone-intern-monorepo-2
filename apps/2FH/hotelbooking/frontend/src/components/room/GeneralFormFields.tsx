'use client';
import { TypePerson, RoomInformation } from '@/generated';
interface GeneralFormProps {
  formData: FormData;
  errors: Record<string, string>;
  onInputChange: (_field: keyof FormData, _value: string | string[]) => void;
}
interface FormData {
  name: string;
  type: string[];
  pricePerNight: string;
  roomInformation: string[];
  bedNumber: number;
}
const TYPE_OPTIONS = [
  { value: TypePerson.Single, label: 'Single' },
  { value: TypePerson.Double, label: 'Double' },
  { value: TypePerson.Triple, label: 'Triple' },
  { value: TypePerson.Quad, label: 'Quad' },
  { value: TypePerson.Queen, label: 'Queen' },
  { value: TypePerson.King, label: 'King' },
];
const ROOM_OPTIONS = [
  { value: RoomInformation.PrivateBathroom, label: 'Private Bathroom' },
  { value: RoomInformation.SharedBathroom, label: 'Shared Bathroom' },
  { value: RoomInformation.FreeBottleWater, label: 'Free Bottle Water' },
  { value: RoomInformation.AirConditioner, label: 'Air Conditioner' },
  { value: RoomInformation.Tv, label: 'TV' },
  { value: RoomInformation.Minibar, label: 'Minibar' },
  { value: RoomInformation.FreeWifi, label: 'Free WiFi' },
  { value: RoomInformation.FreeParking, label: 'Free Parking' },
  { value: RoomInformation.Shower, label: 'Shower' },
  { value: RoomInformation.Bathtub, label: 'Bathtub' },
  { value: RoomInformation.HairDryer, label: 'Hair Dryer' },
  { value: RoomInformation.Desk, label: 'Desk' },
  { value: RoomInformation.Elevator, label: 'Elevator' },
];
export const NameField = ({ formData, errors, onInputChange }: GeneralFormProps) => (
  <div>
    <label htmlFor="room-name" className="block text-sm font-medium text-gray-700 mb-1">
      Name
    </label>
    <input
      id="room-name"
      type="text"
      value={formData.name}
      onChange={(e) => onInputChange('name', e.target.value)}
      className={`w-full px-3 py-2 border rounded-md focus:outline-none focus:ring-2 focus:border-transparent ${
        errors.name ? 'border-red-300 focus:ring-red-500' : 'border-gray-300 focus:ring-blue-500'
      }`}
      placeholder="Enter room name"
    />
    {errors.name && <p className="mt-1 text-sm text-red-600">{errors.name}</p>}
  </div>
);
export const TypeField = ({ formData, errors, onInputChange }: GeneralFormProps) => (
  <div>
    <label htmlFor="room-type" className="block text-sm font-medium text-gray-700 mb-1">
      Type
    </label>
    <div className="relative">
      <select
        id="room-type"
        value={formData.type[0] || ''}
        onChange={(e) => onInputChange('type', [e.target.value])}
        className={`w-full px-3 py-2 border rounded-md focus:outline-none focus:ring-2 focus:border-transparent appearance-none bg-white ${
          errors.type ? 'border-red-300 focus:ring-red-500' : 'border-gray-300 focus:ring-blue-500'
        }`}
      >
        <option value="">Select room type</option>
        {TYPE_OPTIONS.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
      <div className="absolute inset-y-0 right-0 flex items-center pr-3 pointer-events-none">
        <svg className="w-4 h-4 text-gray-400" fill="none" stroke="currentColor" viewBox="0 0 24 24">
          <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
        </svg>
      </div>
    </div>
    {errors.type && <p className="mt-1 text-sm text-red-600">{errors.type}</p>}
  </div>
);
export const PriceField = ({ formData, errors, onInputChange }: GeneralFormProps) => (
  <div>
    <label htmlFor="room-price" className="block text-sm font-medium text-gray-700 mb-1">
      Price per night
    </label>
    <input
      id="room-price"
      type="number"
      value={formData.pricePerNight}
      onChange={(e) => onInputChange('pricePerNight', e.target.value)}
      className={`w-full px-3 py-2 border rounded-md focus:outline-none focus:ring-2 focus:border-transparent ${
        errors.pricePerNight ? 'border-red-300 focus:ring-red-500' : 'border-gray-300 focus:ring-blue-500'
      }`}
      placeholder="Enter price"
      min="0"
      step="100"
    />
    {errors.pricePerNight && <p className="mt-1 text-sm text-red-600">{errors.pricePerNight}</p>}
  </div>
);
export const BedNumberField = ({ formData, errors, onInputChange }: GeneralFormProps) => (
  <div>
    <label htmlFor="bed-number" className="block text-sm font-medium text-gray-700 mb-1">
      Bed number
    </label>
    <input
      id="bed-number"
      type="number"
      value={formData.bedNumber ?? ''}
      onChange={(e) => onInputChange('bedNumber', e.target.value)}
      className={`w-full px-3 py-2 border rounded-md focus:outline-none focus:ring-2 focus:border-transparent ${
        errors.bedNumber ? 'border-red-300 focus:ring-red-500' : 'border-gray-300 focus:ring-blue-500'
      }`}
      placeholder="1"
      min="0"
      step="1"
    />
    {errors.bedNumber && <p className="mt-1 text-sm text-red-600">{errors.bedNumber}</p>}
  </div>
);
export const RoomInformationField = ({ formData, errors, onInputChange }: GeneralFormProps) => {
  const handleCheckboxChange = (optionValue: string, checked: boolean) => {
    if (checked) {
      onInputChange('roomInformation', [...formData.roomInformation, optionValue]);
    } else {
      onInputChange(
        'roomInformation',
        formData.roomInformation.filter((item) => item !== optionValue)
      );
    }
  };
  return (
    <div>
      <label htmlFor="room-information" className="block text-sm font-medium text-gray-700 mb-2">
        Room information
      </label>
      <div id="room-information" className={`grid grid-cols-2 gap-2 max-h-32 overflow-y-auto p-2 rounded-md border ${errors.roomInformation ? 'border-red-300 bg-red-50' : 'border-gray-200'}`}>
        {ROOM_OPTIONS.map((option) => (
          <label key={option.value} className="flex items-center space-x-2 text-sm">
            <input
              type="checkbox"
              id={`room-info-${option.value}`}
              checked={formData.roomInformation.includes(option.value)}
              onChange={(e) => handleCheckboxChange(option.value, e.target.checked)}
              className="rounded border-gray-300 text-blue-600 focus:ring-blue-500"
            />
            <span>{option.label}</span>
          </label>
        ))}
      </div>
      {errors.roomInformation && <p className="mt-1 text-sm text-red-600">{errors.roomInformation}</p>}
    </div>
  );
};
