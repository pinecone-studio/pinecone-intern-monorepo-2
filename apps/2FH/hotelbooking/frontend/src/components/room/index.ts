export * from './General';
export * from './Upcoming';
export * from './RoomImage';
export * from './Roomservice';
