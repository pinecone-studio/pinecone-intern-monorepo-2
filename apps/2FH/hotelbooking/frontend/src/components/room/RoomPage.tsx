'use client';

import { useState } from 'react';
import Image from 'next/image';
import { ChevronLeft } from 'lucide-react';
import { useHotelsQuery, useCreateRoomMutation } from '@/generated';
import { General } from './General';
import { Upcoming } from './Upcoming';
import { Roomservice } from './Roomservice';
import { ImageModal } from './ImageModal';
import { SaveHandler } from './SaveHandler';

export const RoomPage = () => {
  const [selectedHotelId, setSelectedHotelId] = useState<string>('');
  const [isImageModalOpen, setIsImageModalOpen] = useState(false);
  const [roomImages, setRoomImages] = useState<string[]>([]);
  const [roomData, setRoomData] = useState({
    general: {
      name: '',
      type: [] as string[],
      pricePerNight: '',
      roomInformation: [] as string[],
      bedNumber: 0,
      status: 'available', // This should match the Status enum value
    },
    services: {
      bathroom: [] as string[],
      accessibility: [] as string[],
      entertainment: [] as string[],
      foodAndDrink: [] as string[],
      other: [] as string[],
      internet: [] as string[],
      bedRoom: [] as string[],
    },
    images: [] as string[],
  });
  const { data: hotelsData, loading: hotelsLoading, error: hotelsError } = useHotelsQuery();
  const [, { loading: createRoomLoading }] = useCreateRoomMutation();
  const handleImageSave = (images: string[]) => {
    console.log('handleImageSave called with images:', images);
    setRoomImages(images);
    setRoomData((prev) => {
      const updated = { ...prev, images };
      return updated;
    });
    setIsImageModalOpen(false);
  };
  const handleOpenImageModal = () => {
    setIsImageModalOpen(true);
  };
  const handleGeneralSave = (data: typeof roomData.general) => {
    setRoomData((prev) => ({ ...prev, general: data }));
  };
  const handleServiceSave = (data: typeof roomData.services) => {
    setRoomData((prev) => ({ ...prev, services: data }));
  };
  return (
    <div className="min-h-screen bg-gray-50 p-6">
      <div className="max-w-7xl mx-auto">
        <div className="flex items-center gap-4 mb-6">
          <button data-cy="Chevron-Left" className="flex items-center justify-center w-8 h-8 border border-gray-300 rounded-md hover:bg-gray-100 transition-colors">
            <ChevronLeft className="w-4 h-4" />
          </button>
        </div>
        <div className="flex-1 mb-6">
          <h1 data-cy="Room-Name" className="text-xl font-semibold text-gray-900 mb-2">
            {roomData.general.name || 'New Room'}
          </h1>
          <input
            type="text"
            value={roomData.general.name}
            onChange={(e) => {
              setRoomData((prev) => ({
                ...prev,
                general: { ...prev.general, name: e.target.value },
              }));
            }}
            placeholder="Enter room name"
            className="w-full max-w-md px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500 text-sm"
          />
        </div>
        <div className="mb-6">
          <label data-cy="Select-Hotel" className="block text-sm font-medium text-gray-700 mb-2">
            Select Hotel
          </label>
          <select
            data-cy="Select-Hotel-Option"
            value={selectedHotelId}
            onChange={(e) => setSelectedHotelId(e.target.value)}
            className="w-full max-w-md px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
            disabled={hotelsLoading}
          >
            <option value="">Select a hotel</option>
            {hotelsData?.hotels.map((hotel) => (
              <option key={hotel.id} value={hotel.id}>
                {hotel.name}
              </option>
            ))}
          </select>
          {hotelsError && <p className="text-red-500 text-sm mt-1">Error loading hotels: {hotelsError.message}</p>}
        </div>
        <div className="grid grid-cols-1 lg:grid-cols-2 gap-6 mb-6">
          <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-6">
            <General onSave={handleGeneralSave} loading={createRoomLoading} _onImageSave={() => undefined} _data={roomData.general} />
          </div>
          <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-6">
            <div className="flex justify-between items-center mb-4">
              <h2 className="text-lg font-semibold text-gray-900">Images</h2>
              <button data-cy="Edit-Images" className="px-4 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition-colors text-sm font-medium" onClick={handleOpenImageModal}>
                Edit
              </button>
            </div>
            {roomImages.length > 0 ? (
              <div className="grid grid-cols-2 gap-4">
                {roomImages.map((image, index) => (
                  <div key={index} className="relative">
                    {image && <Image src={image} alt={`Room image ${index + 1}`} width={200} height={128} className="w-full h-32 object-cover rounded-md" />}
                  </div>
                ))}
              </div>
            ) : (
              <div className="flex flex-col items-center justify-center py-8">
                <div className="w-16 h-16 bg-gray-100 rounded-full flex items-center justify-center mb-4">
                  <svg className="w-8 h-8 text-gray-400" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={2}
                      d="M4 16l4.586-4.586a2 2 0 012.828 0L16 16m-2-2l1.586-1.586a2 2 0 012.828 0L20 14m-6-6h.01M6 20h12a2 2 0 002-2V6a2 2 0 00-2-2H6a2 2 0 00-2 2v12a2 2 0 002 2z"
                    />
                  </svg>
                </div>
                <h3 className="text-base font-medium text-gray-900 mb-2">No Photos Uploaded</h3>
                <p className="text-sm text-gray-500 text-center max-w-xs">Add photos of your rooms, amenities, or property to showcase your hotel.</p>
              </div>
            )}
          </div>
        </div>
        <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-6 mb-6">
          <Upcoming />
        </div>
        <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-6 mb-6">
          <Roomservice onSave={handleServiceSave} loading={createRoomLoading} _data={roomData.services} />
        </div>
        <SaveHandler selectedHotelId={selectedHotelId} roomData={roomData} setRoomData={setRoomData} loading={createRoomLoading} />
        <ImageModal isOpen={isImageModalOpen} onClose={() => setIsImageModalOpen(false)} onSave={handleImageSave} />
      </div>
    </div>
  );
};
