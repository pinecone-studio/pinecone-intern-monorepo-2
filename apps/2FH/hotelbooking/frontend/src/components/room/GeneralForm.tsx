'use client';

import { NameField, TypeField, PriceField, RoomInformationField } from './GeneralFormFields';

interface GeneralFormProps {
  formData: FormData;
  errors: Record<string, string>;
  onInputChange: (_field: keyof FormData, _value: string | string[]) => void;
}

interface FormData {
  name: string;
  type: string[];
  pricePerNight: string;
  roomInformation: string[];
}

export const GeneralForm = (props: GeneralFormProps) => (
  <div className="space-y-4">
    <NameField {...props} />
    <TypeField {...props} />
    <PriceField {...props} />
    <RoomInformationField {...props} />
  </div>
);
