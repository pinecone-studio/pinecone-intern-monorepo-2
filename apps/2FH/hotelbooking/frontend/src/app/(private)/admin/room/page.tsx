'use client';

import { RoomPage } from '@/components/room';

export const dynamic = 'force-dynamic';

const Home = () => {
  return <RoomPage />;
};

export default Home;
