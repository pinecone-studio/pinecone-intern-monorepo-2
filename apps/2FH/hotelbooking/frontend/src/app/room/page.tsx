'use client';

import { RoomPage } from '@/components/room';

const Home = () => {
  return <RoomPage />;
};

export default Home;
