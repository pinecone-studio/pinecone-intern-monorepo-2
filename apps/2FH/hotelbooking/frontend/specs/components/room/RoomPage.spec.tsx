/* eslint-disable  */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { RoomPage } from '@/components/room/RoomPage';
import { useHotelsQuery, useCreateRoomMutation } from '@/generated';

// Mock dependencies
jest.mock('@/generated', () => ({
  useHotelsQuery: jest.fn(),
  useCreateRoomMutation: jest.fn(),
}));

jest.mock('@/components/room/General', () => ({
  General: ({ onSave, loading, _onImageSave, _data }: any) => (
    <div data-testid="general-component">
      <button onClick={() => onSave({ name: 'Test Room', type: ['double'], pricePerNight: '100', roomInformation: ['wifi'], bedNumber: 2, status: 'available' })}>Save General</button>
      <div>Loading: {loading ? 'true' : 'false'}</div>
    </div>
  ),
}));

jest.mock('@/components/room/Upcoming', () => ({
  Upcoming: () => <div data-testid="upcoming-component">Upcoming Component</div>,
}));

jest.mock('@/components/room/Roomservice', () => ({
  Roomservice: ({ onSave, loading, _data }: any) => (
    <div data-testid="roomservice-component">
      <button onClick={() => onSave({ bathroom: ['shower'], accessibility: [], entertainment: ['tv'], foodAndDrink: [], other: [], internet: ['wifi'], bedRoom: ['air_conditioner'] })}>
        Save Services
      </button>
      <div>Loading: {loading ? 'true' : 'false'}</div>
    </div>
  ),
}));

jest.mock('@/components/room/ImageModal', () => ({
  ImageModal: ({ isOpen, onClose, onSave }: any) =>
    isOpen ? (
      <div data-testid="image-modal">
        <button onClick={() => onSave(['/image1.jpg', '/image2.jpg'])}>Save Images</button>
        <button onClick={onClose}>Close Modal</button>
      </div>
    ) : null,
}));

jest.mock('@/components/room/SaveHandler', () => ({
  SaveHandler: ({ selectedHotelId, roomData, loading }: any) => (
    <div data-testid="save-handler">
      <div>Selected Hotel: {selectedHotelId}</div>
      <div>Room Name: {roomData.general.name}</div>
      <div>Loading: {loading ? 'true' : 'false'}</div>
    </div>
  ),
}));

const mockUseHotelsQuery = useHotelsQuery as jest.MockedFunction<typeof useHotelsQuery>;
const mockUseCreateRoomMutation = useCreateRoomMutation as jest.MockedFunction<typeof useCreateRoomMutation>;

describe('RoomPage', () => {
  const mockHotels = [
    { id: 'hotel-1', name: 'Grand Hotel' },
    { id: 'hotel-2', name: 'City Hotel' },
  ];

  const mockCreateRoom = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    mockUseHotelsQuery.mockReturnValue({
      data: { hotels: mockHotels },
      loading: false,
      error: undefined,
    } as any);

    mockUseCreateRoomMutation.mockReturnValue([mockCreateRoom, { loading: false, error: undefined }] as any);
  });

  it('renders room page correctly', () => {
    render(<RoomPage />);

    expect(screen.getByText('New Room')).toBeInTheDocument();
    expect(screen.getByText('Select Hotel')).toBeInTheDocument();
    expect(screen.getByTestId('general-component')).toBeInTheDocument();
    expect(screen.getByTestId('upcoming-component')).toBeInTheDocument();
    expect(screen.getByTestId('roomservice-component')).toBeInTheDocument();
    expect(screen.getByTestId('save-handler')).toBeInTheDocument();
  });

  it('displays room name input', () => {
    render(<RoomPage />);

    const nameInput = screen.getByPlaceholderText('Enter room name');
    expect(nameInput).toBeInTheDocument();
    expect(nameInput).toHaveValue('');
  });

  it('handles room name input change', () => {
    render(<RoomPage />);

    const nameInput = screen.getByPlaceholderText('Enter room name');
    fireEvent.change(nameInput, { target: { value: 'Updated Room Name' } });

    expect(nameInput).toHaveValue('Updated Room Name');
    expect(screen.getByText('Updated Room Name')).toBeInTheDocument();
  });

  it('displays hotel selection dropdown', () => {
    render(<RoomPage />);

    const hotelSelect = screen.getByDisplayValue('Select a hotel');
    expect(hotelSelect).toBeInTheDocument();

    // Check if hotel options are available
    expect(screen.getByText('Grand Hotel')).toBeInTheDocument();
    expect(screen.getByText('City Hotel')).toBeInTheDocument();
  });

  it('handles hotel selection', () => {
    render(<RoomPage />);

    const hotelSelect = screen.getByDisplayValue('Select a hotel');
    fireEvent.change(hotelSelect, { target: { value: 'hotel-1' } });

    expect(hotelSelect).toHaveValue('hotel-1');
    expect(screen.getByText('Selected Hotel: hotel-1')).toBeInTheDocument();
  });

  it('displays loading state for hotels', () => {
    mockUseHotelsQuery.mockReturnValue({
      data: undefined,
      loading: true,
      error: undefined,
    } as any);

    render(<RoomPage />);

    const hotelSelect = screen.getByDisplayValue('Select a hotel');
    expect(hotelSelect).toBeDisabled();
  });

  it('displays hotel loading error', () => {
    const error = new Error('Failed to load hotels');
    mockUseHotelsQuery.mockReturnValue({
      data: undefined,
      loading: false,
      error,
    } as any);

    render(<RoomPage />);

    expect(screen.getByText('Error loading hotels: Failed to load hotels')).toBeInTheDocument();
  });

  it('handles general component save', () => {
    render(<RoomPage />);

    const saveButton = screen.getByText('Save General');
    fireEvent.click(saveButton);

    expect(screen.getByText('Test Room')).toBeInTheDocument();
  });

  it('handles room service component save', () => {
    render(<RoomPage />);

    const saveButton = screen.getByText('Save Services');
    fireEvent.click(saveButton);

    // The component should handle the save without errors
    expect(screen.getByTestId('roomservice-component')).toBeInTheDocument();
  });

  it('handles image modal opening and closing', () => {
    render(<RoomPage />);

    const editImagesButton = screen.getByText('Edit');
    fireEvent.click(editImagesButton);

    expect(screen.getByTestId('image-modal')).toBeInTheDocument();

    const closeButton = screen.getByText('Close Modal');
    fireEvent.click(closeButton);

    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();
  });

  it('handles image modal save', () => {
    render(<RoomPage />);

    const editImagesButton = screen.getByText('Edit');
    fireEvent.click(editImagesButton);

    const saveImagesButton = screen.getByText('Save Images');
    fireEvent.click(saveImagesButton);

    // Should close modal and update images
    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();
  });

  it('displays no images state initially', () => {
    render(<RoomPage />);

    expect(screen.getByText('No Photos Uploaded')).toBeInTheDocument();
    expect(screen.getByText('Add photos of your rooms, amenities, or property to showcase your hotel.')).toBeInTheDocument();
  });

  it('displays images when available', () => {
    render(<RoomPage />);

    // Open image modal and save images
    const editImagesButton = screen.getByText('Edit');
    fireEvent.click(editImagesButton);

    const saveImagesButton = screen.getByText('Save Images');
    fireEvent.click(saveImagesButton);

    // Should now display the images
    expect(screen.getByAltText('Room image 1')).toBeInTheDocument();
    expect(screen.getByAltText('Room image 2')).toBeInTheDocument();
  });

  it('passes loading state to components', () => {
    mockUseCreateRoomMutation.mockReturnValue([mockCreateRoom, { loading: true, error: undefined }] as any);

    render(<RoomPage />);

    // Check that loading state is passed to components
    const loadingElements = screen.getAllByText('Loading: true');
    expect(loadingElements.length).toBeGreaterThan(0);
  });

  it('passes loading state to save handler', () => {
    mockUseCreateRoomMutation.mockReturnValue([mockCreateRoom, { loading: true, error: undefined }] as any);

    render(<RoomPage />);

    // Check that loading state is passed to save handler
    const loadingElements = screen.getAllByText('Loading: true');
    expect(loadingElements.length).toBeGreaterThan(0);
  });

  it('handles room name input with special characters', () => {
    render(<RoomPage />);

    const nameInput = screen.getByPlaceholderText('Enter room name');
    fireEvent.change(nameInput, { target: { value: 'Room & Suite @#$%' } });

    expect(nameInput).toHaveValue('Room & Suite @#$%');
    expect(screen.getByText('Room & Suite @#$%')).toBeInTheDocument();
  });

  it('handles room name input with empty string', () => {
    render(<RoomPage />);

    const nameInput = screen.getByPlaceholderText('Enter room name');
    fireEvent.change(nameInput, { target: { value: '' } });

    expect(nameInput).toHaveValue('');
    expect(screen.getByText('New Room')).toBeInTheDocument(); // Should show default
  });

  it('handles room name input with whitespace only', () => {
    render(<RoomPage />);

    const nameInput = screen.getByPlaceholderText('Enter room name');
    fireEvent.change(nameInput, { target: { value: '   ' } });

    expect(nameInput).toHaveValue('   ');
    // Check that the whitespace value is displayed in the save handler
    expect(
      screen.getByText((content, element) => {
        return element?.textContent === 'Room Name:    ';
      })
    ).toBeInTheDocument();
  });

  it('handles room name input with very long text', () => {
    render(<RoomPage />);

    const longName = 'A'.repeat(1000);
    const nameInput = screen.getByPlaceholderText('Enter room name');
    fireEvent.change(nameInput, { target: { value: longName } });

    expect(nameInput).toHaveValue(longName);
    expect(screen.getByText(longName)).toBeInTheDocument();
  });

  it('handles hotel selection with empty value', () => {
    render(<RoomPage />);

    const hotelSelect = screen.getByDisplayValue('Select a hotel');
    fireEvent.change(hotelSelect, { target: { value: '' } });

    expect(hotelSelect).toHaveValue('');
    expect(screen.getByText('Selected Hotel:')).toBeInTheDocument();
  });

  it('handles multiple image modal operations', () => {
    render(<RoomPage />);

    // Open modal first time
    const editImagesButton = screen.getByText('Edit');
    fireEvent.click(editImagesButton);
    expect(screen.getByTestId('image-modal')).toBeInTheDocument();

    // Close modal
    const closeButton = screen.getByText('Close Modal');
    fireEvent.click(closeButton);
    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();

    // Open modal second time
    fireEvent.click(editImagesButton);
    expect(screen.getByTestId('image-modal')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<RoomPage />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
