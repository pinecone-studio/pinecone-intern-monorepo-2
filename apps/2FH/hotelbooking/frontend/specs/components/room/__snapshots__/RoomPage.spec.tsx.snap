// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomPage matches snapshot 1`] = `
<div
  class="min-h-screen bg-gray-50 p-6"
>
  <div
    class="max-w-7xl mx-auto"
  >
    <div
      class="flex items-center gap-4 mb-6"
    >
      <button
        class="flex items-center justify-center w-8 h-8 border border-gray-300 rounded-md hover:bg-gray-100 transition-colors"
        data-cy="Chevron-Left"
      >
        <svg
          class="lucide lucide-chevron-left w-4 h-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m15 18-6-6 6-6"
          />
        </svg>
      </button>
    </div>
    <div
      class="flex-1 mb-6"
    >
      <h1
        class="text-xl font-semibold text-gray-900 mb-2"
        data-cy="Room-Name"
      >
        New Room
      </h1>
      <input
        class="w-full max-w-md px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500 text-sm"
        placeholder="Enter room name"
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-6"
    >
      <label
        class="block text-sm font-medium text-gray-700 mb-2"
        data-cy="Select-Hotel"
      >
        Select Hotel
      </label>
      <select
        class="w-full max-w-md px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
        data-cy="Select-Hotel-Option"
      >
        <option
          value=""
        >
          Select a hotel
        </option>
        <option
          value="hotel-1"
        >
          Grand Hotel
        </option>
        <option
          value="hotel-2"
        >
          City Hotel
        </option>
      </select>
    </div>
    <div
      class="grid grid-cols-1 lg:grid-cols-2 gap-6 mb-6"
    >
      <div
        class="bg-white rounded-lg shadow-sm border border-gray-200 p-6"
      >
        <div
          data-testid="general-component"
        >
          <button>
            Save General
          </button>
          <div>
            Loading: 
            false
          </div>
        </div>
      </div>
      <div
        class="bg-white rounded-lg shadow-sm border border-gray-200 p-6"
      >
        <div
          class="flex justify-between items-center mb-4"
        >
          <h2
            class="text-lg font-semibold text-gray-900"
          >
            Images
          </h2>
          <button
            class="px-4 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition-colors text-sm font-medium"
            data-cy="Edit-Images"
          >
            Edit
          </button>
        </div>
        <div
          class="flex flex-col items-center justify-center py-8"
        >
          <div
            class="w-16 h-16 bg-gray-100 rounded-full flex items-center justify-center mb-4"
          >
            <svg
              class="w-8 h-8 text-gray-400"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
            >
              <path
                d="M4 16l4.586-4.586a2 2 0 012.828 0L16 16m-2-2l1.586-1.586a2 2 0 012.828 0L20 14m-6-6h.01M6 20h12a2 2 0 002-2V6a2 2 0 00-2-2H6a2 2 0 00-2 2v12a2 2 0 002 2z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </div>
          <h3
            class="text-base font-medium text-gray-900 mb-2"
          >
            No Photos Uploaded
          </h3>
          <p
            class="text-sm text-gray-500 text-center max-w-xs"
          >
            Add photos of your rooms, amenities, or property to showcase your hotel.
          </p>
        </div>
      </div>
    </div>
    <div
      class="bg-white rounded-lg shadow-sm border border-gray-200 p-6 mb-6"
    >
      <div
        data-testid="upcoming-component"
      >
        Upcoming Component
      </div>
    </div>
    <div
      class="bg-white rounded-lg shadow-sm border border-gray-200 p-6 mb-6"
    >
      <div
        data-testid="roomservice-component"
      >
        <button>
          Save Services
        </button>
        <div>
          Loading: 
          false
        </div>
      </div>
    </div>
    <div
      data-testid="save-handler"
    >
      <div>
        Selected Hotel: 
      </div>
      <div>
        Room Name: 
      </div>
      <div>
        Loading: 
        false
      </div>
    </div>
  </div>
</div>
`;
